designer/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/ActionService.scala

```diff
@@ -181,9 +190,62 @@ class ActionService(
       dbioRunner.runInTransaction(actionRepository.withLockedTable(dbioAction))
     }
 
+    private def fetchAndUpdateScenarioIfNeeded(processId: ProcessIdWithName)(implicit user: LoggedUser) = {
+      scenarioGraphSource match {
+        case LatestVersion =>
+          processRepository.fetchLatestProcessDetailsForProcessId[ScenarioDetailsShape](processId.id)
+        case FromGraph(scenarioGraph, scenarioLabels, baseScenarioVersionId) =>
+          processRepository
+            .fetchProcessDetailsForId[CanonicalProcess](processId.id, baseScenarioVersionId)
+            .flatMap {
+              case Some(scenario) if CanonicalProcessConverter.toScenarioGraph(scenario.json) == scenarioGraph =>
+                val scenarioWithTargetShape =
+                  scenario.mapScenario(ScenarioShapeFetchStrategy.convertToTargetShape[ScenarioDetailsShape])
+                DBIOAction.successful(Some(scenarioWithTargetShape))
+              case Some(scenario) =>
+                updateScenario(processId, scenarioGraph, scenarioLabels, scenario)
+              case None =>
+                DBIOAction.successful[Option[ScenarioWithDetailsEntity[ScenarioDetailsShape]]](None)
+            }
+      }
+    }
+
+    private def updateScenario(
+        processId: ProcessIdWithName,
+        scenarioGraph: ScenarioGraph,
+        scenarioLabels: Option[List[ProcessingType]],
+        scenario: ScenarioWithDetailsEntity[CanonicalProcess]
+    )(implicit user: LoggedUser) = {
+      val parameters = scenarioParametersServiceProvider.combined
+        .getParametersWithReadPermissionUnsafe(scenario.processingType)
+      val scenarioDetails = ScenarioWithDetailsConversions
+        .fromEntityIgnoringGraphAndValidationResult(scenario, parameters)
+      val updateCommand = UpdateScenarioCommand(scenarioGraph, None, scenarioLabels)
+      processService
+        .updateProcessDBIO(processId, scenarioDetails, updateCommand)
+        .flatMap(response =>
+          response.processResponse match {
+            case Some(value) =>
+              logger.debug(
+                s"Scenario: ${value.processName.value} updated with automatic update." +
+                  s" New version is: ${value.versionId.value}"
+              )
+              processRepository.fetchProcessDetailsForId[ScenarioDetailsShape](processId.id, value.versionId)
```

Comment:
Can we avoid fetching details that were already updated by us? Can't we just do some `scenario.copy(...)` ?